lib/os/onoff.c

```diff
@@ -650,3 +661,255 @@ int onoff_sync_finalize(struct onoff_sync_service *srv,
 
 	return rv;
 }
+
+#if CONFIG_ONOFF_DELAYED
+
+static void timeout_handler(struct _timeout *t);
+
+static inline struct onoff_delayed_client *get_delayed_cli(sys_snode_t *node)
+{
+	struct onoff_client *cli =
+			CONTAINER_OF(node, struct onoff_client, node);
+
+	 return CONTAINER_OF(cli, struct onoff_delayed_client, cli);
+}
+
+/**
+ * @brief Convert input timeout to an absolute timeout.
+ *
+ * Input timeout can be relative or absolute. It is reduced by the offset.
+ *
+ * @param timeout output timeout.
+ * @param t input timeout as provided by the user.
+ * @param offset offset given in system ticks.
+ *
+ * @retval true if request is instant.
+ * @retval false if request is in the future.
+ */
+static bool get_abs_time(k_timeout_t *timeout, k_timeout_t t, uint32_t offset)
+{
+	if (K_TIMEOUT_EQ(t, K_NO_WAIT)) {
+		return true;
+	}
+
+	if (Z_TICK_ABS(t.ticks) < 0) {
+		if (t.ticks < offset) {
+			return true;
+		}
+
+		timeout->ticks =
+			Z_TICK_ABS((z_tick_get() + t.ticks - offset));
+	} else {
+		timeout->ticks =
+			Z_TICK_ABS(Z_TICK_ABS(t.ticks) - offset);
+	}
+
+	return false;
+}
+
+static inline sys_snode_t *sortlist_head_peek(sys_slist_t *list)
+{
+	return sys_slist_peek_head(list);
+}
+
+static inline sys_snode_t *sortlist_get(sys_slist_t *list)
+{
+	return sys_slist_get(list);
+}
+
+static inline bool sortlist_remove(sys_slist_t *list, sys_snode_t *node)
+{
+	return sys_slist_find_and_remove(list, node);
+}
+
+/** @brief Compare to absolute tick values as represented in k_timeout_t.
+ *
+ * @param t0 First absolute tick value.
+ * @param t1 Second absolute tick value.
+ *
+ * @retval true if @p t0 is equal or later than @t1
+ * @retval false if @p t1 is later than @p t0
+ */
+static inline bool abs_tick_compare(int64_t t0, int64_t t1)
```

Comment:
What I meant in my original comment was that if we're going to have API to compare timeouts it belongs at the same level as `K_TIMEOUT_EQ`.  So perhaps `K_TIMEOUT_REL_LT(a,b)` and `K_TIMEOUT_ABS_LT(a,b)`, since comparing a relative and absolute timeout is not trivial. 

And it should operate on timeouts, not `int64_t` (which is not necessarily `k_ticks_t`).